jaxrs/src/main/java/com/facebook/airlift/jaxrs/JerseyAuthorizationFilter.java

```diff
@@ -0,0 +1,110 @@
+import javax.ws.rs.ext.Provider;
+
+import java.security.Principal;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.lang.String.format;
+
+@Provider
+public class JerseyAuthorizationFilter
+        implements ContainerRequestFilter
+{
+    private final Authorizer authorizer;
+    private final HttpServerConfig httpServerConfig;
+
+    @Context
+    private ResourceInfo resourceInfo;
+
+    @Inject
+    public JerseyAuthorizationFilter(Authorizer authorizer, HttpServerConfig httpServerConfig)
+    {
+        this.authorizer = authorizer;
+        this.httpServerConfig = httpServerConfig;
+    }
+
+    @Override
+    public void filter(ContainerRequestContext request)
+    {
+        Principal principal = request.getSecurityContext().getUserPrincipal();
+        if (principal == null) {
+            request.abortWith(Response.status(Response.Status.FORBIDDEN)
+                    .entity("Request principal is missing.")
+                    .build());
+            return;
+        }
+
+        Optional<Set<String>> allowedRoles = getAllowedRoles();
+        if (isEndpointUnmarked(allowedRoles)) {
+            switch (httpServerConfig.getDefaultAuthorizationPolicy()) {
+                case ALLOW:
+                    return;
+                case DENY:
+                    request.abortWith(Response.status(Response.Status.FORBIDDEN)
+                            .entity(format("Principal %s is not allowed to access the resource. Reason: denied by default policy",
+                                    principal.getName()))
+                            .build());
+                    return;
+                case DEFAULT_ROLES:
+                    allowedRoles = Optional.of(httpServerConfig.getDefaultAllowedRoles());
+                    break;
+            }
+        }
+
+        AuthorizationResult result = authorizer.authorize(principal, allowedRoles.get());
+        if (result == null || !result.isAllowed()) {
+            request.abortWith(Response.status(Response.Status.FORBIDDEN)
+                    .entity(format("Principal %s is not allowed to access the resource. Reason: %s",
+                            principal.getName(),
+                            result.getReason().orElse("")))
+                    .build());
+        }
+    }
+
+    private Optional<Set<String>> getAllowedRoles()
+    {
+        if (resourceInfo.getResourceMethod().isAnnotationPresent(RolesAllowed.class)) {
+            return Optional.of(ImmutableSet.copyOf(resourceInfo.getResourceMethod().getAnnotation(RolesAllowed.class).value()));
+        }
+        else if (resourceInfo.getResourceClass().isAnnotationPresent(RolesAllowed.class)) {
+            return Optional.of(ImmutableSet.copyOf(resourceInfo.getResourceClass().getAnnotation(RolesAllowed.class).value()));
+        }
+        else {
+            return Optional.empty();
+        }
+    }
+
+    private boolean isEndpointUnmarked(Optional<Set<String>> allowedRoles)
```

Comment:
nit. you can inline this single line method.

Reply:
Maybe this method will make it more readable? Since `!allowedRoles.isPresent()` itself looks doesn't make much sense without further explanations.

Reply:
I agree with @ajaygeorge here - I feel that `!allowedRoles.isPresent()` implies that it is not marked.